Authentication: Fix immediate update_connection_password to use username

liorsve opened the pull request:
<!--
Thanks for contributing to Valkey GLIDE!

Please make sure you are aware of our contributing guidelines [available
here](https://github.com/valkey-io/valkey-glide/blob/main/CONTRIBUTING.md)

-->

### Issue link

This Pull Request is linked to issue: https://github.com/valkey-io/valkey-glide/issues/3289

This PR fixes the update connection password to add the username as arg in `update_connection_password(immediateAuth = True)`  (i.e. send "AUTH [username] [password]")  in case username was configured in ServerCredentials during client creation.
Currently, only "AUTH [password]" is sent.

Additionally, it fixes the implementation of `update_connection_password` in CMD - previously, the updated password would be saved only in the multiplexed_connection struct, and then upon disconnection the authentication would still be with the old password. Now, the password is updated in the `ReconnectingConnection` struct which is used in reconnecitons.

NOTES- 
- Deleting the user with `delete_acl_username_and_password` is triggering a disconnect from clients authenticated with that username. 
- Unlike the previous tests, calling update_connection_password with non-immediate authentication **when disconnected from server** (for example after deleting the user or actively killing the connection) is supposed to succeed both for CME and CMD. That's because this is an internal operation and doesn't require an active connection. The difference is, when calling with immediate auth, CME can reconnect with that supplied password while CMD tries to reconnect with the initial password and thus won't succeed. 

### Checklist

Before submitting the PR make sure the following are checked:

-   [x] This Pull Request is related to one issue.
-   [x] Commit message has a detailed description of what changed and why.
-   [x] Tests are added or updated.
-   [x] CHANGELOG.md and documentation files are updated.
-   [x] Destination branch is correct - main or release
-   [x] Create merge commit if merging release branch into main, squash otherwise.

Yury-Fridlyand commented on glide-core/src/client/mod.rs:
```diff
@@ -525,6 +525,16 @@ impl Client {
+                        }
+                    }
+                    ClientWrapper::Standalone(_) => {
+                        //TODO: get username and add it as arg in standalone
```
Do you plan to implement this in the current PR?
If no, create a task

liorsve replied:
Implemented in this PR

barshaul commented on node/tests/AuthTest.test.ts:
```diff
@@ -71,6 +75,29 @@ describe("Auth tests", () => {
+        client: BaseClient,
+        username: string,
+    ) {
+        return await client.customCommand(["ACL", "DELUSER", username]);
```
minor:
```suggestion
       await client.customCommand(["ACL", "DELUSER", username])
```

barshaul commented on node/tests/AuthTest.test.ts:
```diff
@@ -71,6 +75,29 @@ describe("Auth tests", () => {
     ): { host: string; port: number }[] =>
         addresses.map(([host, port]) => ({ host, port }));
 
+    async function setNewAclUsernameWithPassword(
```
This and the deleteAclUsernameAndPassword functions should handle the result of the requests  - if the command fails you should catch it and panic, otherwise we won't know why the test is failing

liorsve replied:
I added an assertion everywhere besides `delete_acl_username_and_password` in the python tests - there the acl user is set only for the acl specific tests that I added, so asserting that a client was deleted fails on the non acl-user tests. So instead, in python I had the delete user function return the command's responds (1 if a user was deleted, 0 if not) and i'm asserting inside the relevant tests themselves that it's returning 1 when it should.

liorsve commented:
Changes - 
- addressed previous comments about the CME changes.
- added CMD fix in core + CMD tests.
- In tests, I had to introduce some sleep wherever a server disconnection + password change was made, otherwise they were flaky. I encountered the same timing issue that was talked about here https://github.com/valkey-io/valkey-glide/pull/3311, where the client sometimes wouldn't have enough time to try and reconnect, probably because it takes some time for the new password to register in the server as well, and if the two are not well-synchronized in their timing it results in 
a connectionError. Introducing some sleep after the disconnection seems to have resolved it, so for now I re-added the test that was removed in this PR.

@barshaul @ikolomi @avifenesh

barshaul commented on node/tests/AuthTest.test.ts:
```diff
@@ -291,39 +382,221 @@ describe("Auth tests", () => {
+                    protocol,
+                    clusterMode,
+                );
+            });
```
nit: missing newline

barshaul commented on python/python/tests/conftest.py:
```diff
@@ -239,6 +239,36 @@ async def management_client(
+        request,
+        cluster_mode,
+        protocol=protocol,
+        request_timeout=5000,
```
is it needed for the ACL tests? I think you can remove this and use the default

barshaul commented on python/python/tests/conftest.py:
```diff
@@ -330,6 +364,37 @@ async def config_set_new_password(client: TGlideClient, password):
+        return await client.custom_command(
+            ["ACL", "DELUSER", username], route=AllNodes()
+        )
+
```
else: raise error

barshaul commented on python/python/tests/conftest.py:
```diff
@@ -330,6 +364,37 @@ async def config_set_new_password(client: TGlideClient, password):
+        raise RuntimeError(f"Failed to set ACL user: {e}")
+
+
+async def delete_acl_username_and_password(client: TGlideClient, username):
```
missing typing for username and password across the file

barshaul replied:
move this function to python/python/tests/utils/utils.py

barshaul commented on python/python/tests/conftest.py:
```diff
@@ -330,6 +364,37 @@ async def config_set_new_password(client: TGlideClient, password):
         await client.config_set({"requirepass": password}, route=AllNodes())
 
 
+async def set_new_acl_username_with_password(client: TGlideClient, username, password):
```
move this function to python/python/tests/utils/utils.py

barshaul reviewed (changes requested):
Awesome job! added comments inline. 
What about Java's tests?

liorsve commented on node/tests/AuthTest.test.ts:
```diff
@@ -102,22 +134,43 @@ describe("Auth tests", () => {
-        const isStandaloneMode = configOverrides?.addresses?.length === 1;
+        let isStandaloneMode;
+
+        if (configOverrides?.addresses) {
+            isStandaloneMode =
+                configOverrides.addresses.length === 1 ? false : true;
+        } else {
+            isStandaloneMode = !cluster_mode;
+        }
```
It's not used here, but in theory you could provide in the configOverrides an external address to a different server than the one that was configured in the `beforeAll()` part, and then you would want only the appropriate test to run for the type of addresses provided. it's a leftover from the how the tests were before, but now that i'm thinking about it it's both not correct (addresses.length could be > 1 if replicas provided) and also, different server could be just configured in the cli if one wanted to, so it's redundant.  So i'll remove it anyway

liorsve commented on glide-core/src/client/standalone_client.rs:
```diff
@@ -607,15 +607,18 @@ impl StandaloneClient {
+        Ok(Value::Okay)
+    }
+
+    pub async fn get_username(&self) -> Option<String> {
```
No reason, I think a leftover from a prev version where the inner get_username function was async. Anyways made it sync

barshaul commented on glide-core/src/client/mod.rs:
```diff
@@ -525,11 +525,27 @@ impl Client {
+                Ok(Value::SimpleString(username)) => Some(username),
+                Ok(Value::Nil) => None,
+                Ok(other) => panic!("Expected SimpleString, got: {:?}", other),
+                Err(e) => panic!("Got error trying to get username: {:?}", e),
```
We don't want to panic if get_username fails, as it would crash the whole client. Instead, change get_username to return Result<Option<String>> and return the error if we get one.

barshaul commented on glide-core/src/client/mod.rs:
```diff
@@ -525,11 +525,27 @@ impl Client {
+            ClientWrapper::Cluster { client } => match client.get_username().await {
+                Ok(Value::SimpleString(username)) => Some(username),
+                Ok(Value::Nil) => None,
+                Ok(other) => panic!("Expected SimpleString, got: {:?}", other),
```
```suggestion
                Ok(other) => unreachable!("Expected SimpleString or Nil, got: {:?}", other),
```

barshaul commented on glide-core/src/client/reconnecting_connection.rs:
```diff
@@ -119,7 +120,14 @@ async fn create_connection(
+        let guard = connection_backend
+            .connection_info
+            .read()
+            .expect("READ_LOCK_ERR");
```
you should define consts and use them:
const WRITE_LOCK_ERR: &str = "Failed to acquire the write lock";
const READ_LOCK_ERR: &str = "Failed to acquire the read lock";

barshaul commented on glide-core/src/client/reconnecting_connection.rs:
```diff
@@ -119,7 +120,14 @@ async fn create_connection(
     connection_timeout: Duration,
 ) -> Result<ReconnectingConnection, (ReconnectingConnection, RedisError)> {
-    let client = &connection_backend.connection_info;
+    let client = {
```
rename client to conn_info

barshaul commented on glide-core/src/client/reconnecting_connection.rs:
```diff
@@ -204,6 +213,13 @@ fn internal_retry_iterator() -> impl Iterator<Item = Duration> {
+impl ConnectionBackend {
+    /// Returns a read-only reference to the client's connection information
+    fn get_backend_client(&self) -> RwLockReadGuard<'_, redis::Client> {
+        self.connection_info.read().expect("READ_LOCK_ERR")
```
same regarding the error msg

barshaul commented on glide-core/src/client/reconnecting_connection.rs:
```diff
@@ -363,4 +383,26 @@ impl ReconnectingConnection {
+            .backend
+            .connection_info
+            .write()
+            .expect("WRITE_LOCK_ERR");
```
same reg. excpect err msg

barshaul commented on glide-core/src/client/reconnecting_connection.rs:
```diff
@@ -363,4 +383,26 @@ impl ReconnectingConnection {
+            .backend
+            .connection_info
+            .read()
+            .expect("READ_LOCK_ERR");
```
same

barshaul commented on glide-core/src/client/reconnecting_connection.rs:
```diff
@@ -363,4 +383,26 @@ impl ReconnectingConnection {
+        let client = self
+            .inner
+            .backend
+            .connection_info
```
why don't you use get_backend_client here?

liorsve replied:
Not sure 🤷‍♀️ fixed

barshaul commented on glide-core/src/client/standalone_client.rs:
```diff
@@ -604,18 +604,23 @@ impl StandaloneClient {
     }
 
-    /// Update the password used to authenticate with the servers.
+    /// Update the password used to authenticate with the server.
```
```suggestion
    /// Update the password used to authenticate with the servers.
```

barshaul commented on java/integTest/src/test/java/glide/cluster/ClusterClientTests.java:
```diff
@@ -278,4 +280,177 @@ public void test_replace_password_immediateAuth_wrong_password() {
+            Thread.sleep(1000);
+
+            // Validate client reconnected succsessfuly
+            assertNotNull(testClient.info().get());
```
If you deleted the ACL user and haven't updated the client yet, how does it work? shouldn't we get an error here?

liorsve replied:
I have updated the client's password, 4 lines before

barshaul reviewed (changes requested):
See comments inline

barshaul commented on java/integTest/src/test/java/glide/TestUtilities.java:
```diff
@@ -452,4 +453,104 @@ public static String getServerVersion(@NonNull final BaseClient client) {
+                                        "SETUSER",
+                                        username,
+                                        "on",
+                                        "allkeys",
+                                        "+get",
+                                        "+cluster",
+                                        "+ping",
+                                        "+info",
+                                        "+client",
```
change to all commands

barshaul commented on java/integTest/src/test/java/glide/TestUtilities.java:
```diff
@@ -452,4 +453,104 @@ public static String getServerVersion(@NonNull final BaseClient client) {
+                                    new String[] {
+                                        "ACL",
+                                        "SETUSER",
+                                        username,
```
change to all commands

barshaul commented on glide-core/src/client/mod.rs:
```diff
@@ -525,11 +525,31 @@ impl Client {
+            ClientWrapper::Cluster { client } => match client.get_username().await {
+                Ok(Value::SimpleString(username)) => Ok(Some(username)),
+                Ok(Value::Nil) => Ok(None),
+                Ok(other) => unreachable!("Expected SimpleString or Nil, got: {:?}", other),
```
don't use unreachable as it would panic and crash the whole client. Instead, return:
RedisError::from((
                ErrorKind::ClientError,
                "Expected SimpleString or Nil, got: {other:?}"
            ))

barshaul commented on glide-core/src/client/mod.rs:
```diff
@@ -525,11 +525,31 @@ impl Client {
                     Some(ResponsePolicy::AllSucceeded),
                 ));
                 let mut cmd = redis::cmd("AUTH");
+                if let Ok(Some(username)) = self.get_username().await {
```
don't ignore errors, instead, return the error to the caller by using '?':
```
if let Some(username) = self.get_username().await? {

barshaul reviewed (approved):
Fix last comments, squash the commits & merge
